File: mkdocs/snippets/devbook/websockets/listen-transaction-flow.py
import asyncio
import json
import os
import urllib.request

from symbolchain.CryptoTypes import PrivateKey
from symbolchain.facade.SymbolFacade import SymbolFacade
from symbolchain.sc import Amount
from symbolchain.symbol.Network import NetworkTimestamp
from websockets import connect

NODE_URL = os.getenv('NODE_URL', 'https://reference.symboltest.net:3001')
WS_URL = NODE_URL.replace('http', 'ws', 1) + '/ws'
print(f'Using node {NODE_URL}')
# [>step-1]
MONITOR_ADDRESS = os.getenv(
	'MONITOR_ADDRESS',
	'TCHBDENCLKEBILBPWP3JPB2XNY64OE7PYHHE32I'
)
print(f'Monitoring address: {MONITOR_ADDRESS}')

SIGNER_PRIVATE_KEY = os.getenv(
	'SIGNER_PRIVATE_KEY',
	'0000000000000000000000000000000000000000000000000000000000000000'
)
facade = SymbolFacade('testnet')
signer_key_pair = SymbolFacade.KeyPair(PrivateKey(SIGNER_PRIVATE_KEY))
# [<step-1]

async def main():
	async with connect(WS_URL) as websocket: # [>step-2]
		# Connect to WebSocket
		response = json.loads(await websocket.recv())
		uid = response['uid']
		print(f'Connected to {WS_URL} with uid {uid}')
	# [<step-2]
		# Subscribe to transaction channels [>step-3]
		channels = [
			f'unconfirmedAdded/{MONITOR_ADDRESS}',
			f'unconfirmedRemoved/{MONITOR_ADDRESS}',
			f'confirmedAdded/{MONITOR_ADDRESS}',
		]
		for channel in channels:
			await websocket.send(json.dumps(
				{'uid': uid, 'subscribe': channel}
			))
			name = channel.split('/')[0]
			print(f'Subscribed to {name} channel')
		# [<step-3]
		# Build and announce a transfer transaction [>step-4]
		with urllib.request.urlopen(f'{NODE_URL}/node/time') as resp:
			time_json = json.loads(resp.read().decode())
			timestamp = NetworkTimestamp(int(
				time_json['communicationTimestamps']['receiveTimestamp']))

		with urllib.request.urlopen(
			f'{NODE_URL}/network/fees/transaction'
		) as resp:
			fee_json = json.loads(resp.read().decode())
			fee_mult = max(
				fee_json['medianFeeMultiplier'],
				fee_json['minFeeMultiplier'])

		transaction = facade.transaction_factory.create({
			'type': 'transfer_transaction_v1',
			'signer_public_key': signer_key_pair.public_key,
			'deadline': timestamp.add_hours(2).timestamp,
			'recipient_address': MONITOR_ADDRESS,
		})
		transaction.fee = Amount(fee_mult * transaction.size)

		signature = facade.sign_transaction(signer_key_pair, transaction)
		json_payload = facade.transaction_factory.attach_signature(
			transaction, signature)
		transaction_hash = str(facade.hash_transaction(transaction)) # [<step-4]
		# [>step-5]
		announce_request = urllib.request.Request(
			f'{NODE_URL}/transactions',
			data=json_payload.encode(),
			headers={'Content-Type': 'application/json'},
			method='PUT'
		)
		with urllib.request.urlopen(announce_request) as resp:
			resp.read()
		print(f'Announced transaction {transaction_hash[:16]}...')

		# Wait for confirmation via WebSocket
		async for raw_message in websocket:
			message = json.loads(raw_message)
			topic = message['topic']
			message_hash = message['data']['meta']['hash']
			name = topic.split('/')[0]
			print(f'{name}: hash={message_hash[:16]}...')

			if (name == 'confirmedAdded'
					and message_hash == transaction_hash):
				print(f'Transaction {transaction_hash[:16]}... confirmed')
				break
		# [<step-5]
		# Unsubscribe before closing [>step-6]
		for channel in channels:
			await websocket.send(json.dumps({
				'uid': uid,
				'unsubscribe': channel
			}))
		print('Unsubscribed from all channels')
		# [<step-6]

try:
	asyncio.run(main())
except Exception as error:
	print(error)

File: mkdocs/snippets/devbook/websockets/listen_transaction_flow.mjs
import { PrivateKey } from 'symbol-sdk';
import { NetworkTimestamp, SymbolFacade, models } from 'symbol-sdk/symbol';

const NODE_URL = process.env.NODE_URL ||
	'https://reference.symboltest.net:3001';
const WS_URL = `${NODE_URL.replace('http', 'ws')}/ws`;
console.log(`Using node ${NODE_URL}`);
// [>step-1]
const MONITOR_ADDRESS = process.env.MONITOR_ADDRESS ||
	'TCHBDENCLKEBILBPWP3JPB2XNY64OE7PYHHE32I';
console.log(`Monitoring address: ${MONITOR_ADDRESS}`);

const SIGNER_PRIVATE_KEY = process.env.SIGNER_PRIVATE_KEY ||
	'0000000000000000000000000000000000000000000000000000000000000000';
const facade = new SymbolFacade('testnet');
const signerKeyPair = new SymbolFacade.KeyPair(
	new PrivateKey(SIGNER_PRIVATE_KEY));
// [<step-1]
try {
	// Connect to WebSocket [>step-2]
	const websocket = new WebSocket(WS_URL);
	const uid = await new Promise(resolve => {
		websocket.addEventListener('message', event => {
			const message = JSON.parse(event.data);
			resolve(message.uid);
		}, { once: true });
	});
	console.log(`Connected to ${WS_URL} with uid ${uid}`);
	// [<step-2]
	// Subscribe to transaction channels [>step-3]
	const channels = [
		`unconfirmedAdded/${MONITOR_ADDRESS}`,
		`unconfirmedRemoved/${MONITOR_ADDRESS}`,
		`confirmedAdded/${MONITOR_ADDRESS}`
	];
	for (const channel of channels) {
		websocket.send(JSON.stringify({ uid, subscribe: channel }));
		const name = channel.split('/')[0];
		console.log(`Subscribed to ${name} channel`);
	}
	// [<step-3]
	// Build and announce a transfer transaction [>step-4]
	const timeResponse = await fetch(`${NODE_URL}/node/time`);
	const timeJSON = await timeResponse.json();
	const timestamp = new NetworkTimestamp(
		timeJSON.communicationTimestamps.receiveTimestamp);

	const feeResponse = await fetch(
		`${NODE_URL}/network/fees/transaction`);
	const feeJSON = await feeResponse.json();
	const feeMult = Math.max(
		feeJSON.medianFeeMultiplier, feeJSON.minFeeMultiplier);

	const transaction = facade.transactionFactory.create({
		type: 'transfer_transaction_v1',
		signerPublicKey: signerKeyPair.publicKey.toString(),
		deadline: timestamp.addHours(2).timestamp,
		recipientAddress: MONITOR_ADDRESS
	});
	transaction.fee = new models.Amount(feeMult * transaction.size);

	const signature = facade.signTransaction(signerKeyPair, transaction);
	const jsonPayload = facade.transactionFactory.static.attachSignature(
		transaction, signature);
	const transactionHash = facade.hashTransaction(transaction).toString(); // [<step-4]
	// [>step-5]
	const confirmed = new Promise(resolve => {
		websocket.addEventListener('message', event => {
			const message = JSON.parse(event.data);
			const topic = message.topic;
			const messageHash = message.data.meta.hash;
			const name = topic.split('/')[0];
			console.log(
				`${name}: hash=${messageHash.substring(0, 16)}...`);

			if ('confirmedAdded' === name &&
				messageHash === transactionHash) {
				console.log(
					`Transaction ${transactionHash.substring(0, 16)}` +
						'... confirmed');
				resolve();
			}
		});
	});
	await fetch(`${NODE_URL}/transactions`, {
		method: 'PUT',
		headers: { 'Content-Type': 'application/json' },
		body: jsonPayload
	});
	console.log(
		'Announced transaction ' +
		`${transactionHash.substring(0, 16)}...`);

	// Wait for confirmation via WebSocket
	await confirmed;
	// [<step-5]
	// Unsubscribe before closing [>step-6]
	for (const channel of channels)
		websocket.send(JSON.stringify({ uid, unsubscribe: channel }));
	console.log('Unsubscribed from all channels');
	websocket.close(); // [<step-6]
} catch (error) {
	console.error(error);
}
